test/proxy.t.sol

```diff
@@ -52,93 +50,229 @@ contract ProxyTest is DSTest {
         Proxy proxy = Proxy(proxyAddr);
 
-        // using action contract storage should fail
-        bytes memory data = abi.encodeWithSignature("doAdd(uint256,uint256)", address(core), 5,7);
-        bytes memory response = proxy.execute(address(action), data);
+        // use non-existant function call
+        bytes memory data = abi.encodeWithSignature("inlineSubtract(uint256,uint256)", 5,7);
+
+        // set action as a safe target
+        proxy.file("target", address(action));
+
+        // Add this as a user so it can execute
+        proxy.addUser(user);
+
+        // execute action that does not call core contract
+        vm.prank(user);
+        bytes memory response = proxy.userExecute(address(action), data);
     }
 
-    function testFailExecuteNotNFTOwner() public {
+    function testExecuteNotUserFails() public {
         address payable proxyAddr = registry.build();
         Proxy proxy = Proxy(proxyAddr);
 
-        uint accessToken = proxy.accessToken();
-        registry.transferFrom(msg.sender,address(123), accessToken);
+        bytes memory data = abi.encodeWithSignature("inlineAdd(uint256,uint256)", 5,7);
+
+        // set action as a safe target
+        proxy.file("target", address(action));
 
-        // using core
-        bytes memory data = abi.encodeWithSignature("doAdd(address,uint256,uint256)", address(core), 5,7);
+        // Add this as a user so it can execute
+        // proxy.addUser(address(this));
 
-        // should fail because doesn't own accessToken anymore
-        bytes memory  response = proxy.execute(address(action), data);
+        // execute action that does not call core contract
+        vm.expectRevert(bytes("TinlakeProxy/user-not-authorized"));
+        bytes memory response = proxy.userExecute(address(action), data);
     }
 
-    function testExecuteCode() public {
+    function testExecuteNotSafeTargetFails(address user) public {
         address payable proxyAddr = registry.build();
         Proxy proxy = Proxy(proxyAddr);
 
-        bytes memory data = abi.encodeWithSignature("getBytes32AndUint()");
+        bytes memory data = abi.encodeWithSignature("inlineAdd(uint256,uint256)", 5,7);
 
-        //deploy and call the contracts code
-        (, bytes memory response) = proxy.executeByteCode(testCode, data);
+        // set action as a safe target
+        // proxy.file("target", address(action));
 
-        bytes32 response32;
-        uint responseUint;
+        // Add this as a user so it can execute
+        proxy.addUser(user);
 
-        assembly {
-            response32 := mload(add(response, 0x20))
-            responseUint := mload(add(response, 0x40))
-        }
+        // execute action that does not call core contract
+        vm.prank(user);
+        vm.expectRevert(bytes("TinlakeProxy/target-not-authorized"));
+        bytes memory response = proxy.userExecute(address(action), data);
+    }
+
+    function testFailExecuteAccessActionStorage(address user) public {
```

Comment:
Can we use `vm.expectRevert` for this somehow? `testFail` is not ideal as it will also "succeed" if it fails for any other reason that's not the expected failure reason.

Reply:
This one just reverts on an EVM error. No message or anything. I could throw a generic vm.expectRevert() in there but thats not much better. What do you think?

Reply:
Fair enough, let's keep it as is then.